db/crud.go

```diff
@@ -814,6 +814,240 @@ func (db *Database) PutExistingRev(docid string, body Body, docHistory []string,
+	}
+
+	return body, newAttachments, shadowerEcho, updatedExpiry, nil
+}
+
+func (db *Database) getWinningRevAndSetDocFlags(newRevID string, doc *Document) (prevCurrentRev string) {
+	prevCurrentRev = doc.CurrentRev
+	var branched, inConflict bool
+	doc.CurrentRev, branched, inConflict = doc.History.winningRevision()
+	doc.setFlag(channels.Deleted, doc.History[doc.CurrentRev].Deleted)
+	doc.setFlag(channels.Conflict, inConflict)
+	doc.setFlag(channels.Branched, branched)
+	return prevCurrentRev
+}
+
+func (db *Database) setTombstoneTimeIfRequired(doc *Document) {
+	if doc.hasFlag(channels.Deleted) {
+		doc.SyncData.TombstonedAt = time.Now().Unix()
+	} else {
+		doc.SyncData.TombstonedAt = 0
+	}
+}
+
+func (db *Database) storeOldBodyInRevTreeAndUpdateCurrent(doc *Document, prevCurrentRev string, newRevID string, body Body) Body {
+	if doc.CurrentRev != prevCurrentRev && prevCurrentRev != "" && doc.Body() != nil {
+		// Store the doc's previous body into the revision tree:
+		bodyJSON, marshalErr := doc.MarshalBody()
+		if marshalErr != nil {
+			base.WarnfCtx(db.Ctx, base.KeyAll, "Unable to marshal document body for storage in rev tree: %v", marshalErr)
+		}
+		doc.setNonWinningRevisionBody(prevCurrentRev, bodyJSON, db.AllowExternalRevBodyStorage())
+	}
+	// Store the new revision body into the doc:
+	storedBody := doc.setRevisionBody(newRevID, body, db.AllowExternalRevBodyStorage())
+
+	if doc.CurrentRev == newRevID {
+		doc.NewestRev = ""
+		doc.setFlag(channels.Hidden, false)
+	} else {
+		doc.NewestRev = newRevID
+		doc.setFlag(channels.Hidden, true)
+		if doc.CurrentRev != prevCurrentRev {
+			doc.promoteNonWinningRevisionBody(doc.CurrentRev, db.RevisionBodyLoader)
+		}
+	}
+	return storedBody
+}
+
+func (db *Database) runSyncFn(doc *Document, body Body, newRevId string, docid string) (Body, *uint32, string, error, base.Set, channels.AccessMap, channels.AccessMap) {
+	body[BodyId] = doc.ID
+	channelSet, access, roles, syncExpiry, oldBody, err := db.getChannelsAndAccess(doc, body, newRevId)
+	if err != nil {
+		return nil, nil, ``, err, nil, nil, nil
+	}
+	db.checkDocChannelsAndGrantsLimits(docid, channelSet, access, roles)
+	if len(channelSet) > 0 {
+		doc.History[newRevId].Channels = channelSet
+	}
+	return body, syncExpiry, oldBody, nil, channelSet, access, roles
+}
+
+func (db *Database) addAttachments(newAttachments AttachmentData) error {
+	// Need to check and add attachments here to ensure the attachment is within size constraints
+	err := db.setAttachments(newAttachments)
+	if err != nil {
+		if err.Error() == "document value was too large" {
+			err = base.HTTPErrorf(http.StatusRequestEntityTooLarge, "Attachment too large")
+		} else {
+			err = errors.Wrap(err, "Error adding attachment")
+		}
+	}
+	return err
+}
+
+// Sequence processing
+func (db *Database) handleSequences(docSequence uint64, doc *Document, docid string, unusedSequences []uint64) ([]uint64, error, uint64) {
```

Comment:
move `err` to last return parameter, and name variables - can then remove `var err error`